### PR TITLE
Shopping cart: Fix cart key change and reload behavior

### PR DESCRIPTION
**Pull request by sirbrillig**
#### Changes proposed in this Pull Request

This fixes some bugs in the `@automattic/shopping-cart` package. None of the bugs here are causing any problems currently, but they will in the future as the package gains use. Particularly, this fixes the cart reload process where the cart will re-fetch its data from the server when the cart key changes or when manually requested to do so.

It also makes sure that the fake "wpcom credits" products generated by the shopping cart endpoint are stripped out before being passed along to consumers. They are not necessary because credits are available as a separate set of properties on the `ResponseCart` object and they require special exceptions in any code that iterates over products (eg: is the cart empty if it still contains a credits product?). Eventually there's an intent to remove these products from the backend as well, but for now we can at least make sure that frontend consumers don't have to deal with them.

#### Testing instructions

Verify that checkout is unchanged; add a product to your cart and visit checkout. Verify that there are no errors and that the cart product appears. Delete the product and verify that it goes away.

If you wish to test these new modifications themselves, you'll need to test with changes that use these features. https://github.com/Automattic/wp-calypso/pull/47605 requires them, so testing that PR will also validate these changes.

https://github.com/Automattic/wp-calypso/pull/47648 also adds integration tests for this fix.

### CODE REVIEW

File: client/my-sites/checkout/composite-checkout/test/translate-cart.js [hunk 1]
```diff
@@ -818,199 +818,4 @@ describe( 'translateResponseCartToWPCOMCart', function () {
 			} );
 		} );
 	} );
-
-	describe( 'Cart with one plan only (BRL) and a credits line item', function () {
```
Comment (nbloomf) [inline client/my-sites/checkout/composite-checkout/test/translate-cart.js:822]: Is it worth keeping this test and just updating it to work with the credits line item removed?

Comment (sirbrillig) [reply client/my-sites/checkout/composite-checkout/test/translate-cart.js:822]: I think that's what this is: https://github.com/Automattic/wp-calypso/blob/65f7cb7ba5d738b35af072549fcd5af00d8b3a70/client/my-sites/checkout/composite-checkout/test/translate-cart.js#L7

**Review by pottedmeat — COMMENTED** [review]
Do you know how `emptyResponseCart` was triggering mutation detection? This seems like it creates more empty cart objects than before.

**Comment by sirbrillig** [comment]
> Do you know how emptyResponseCart was triggering mutation detection? This seems like it creates more empty cart objects than before.

I'm not sure what you mean by triggering mutation detection, but the reason I wanted to get rid of `emptyResponseCart` is that the object was being exported and then imported in many places; if any of those places had accidentally modified the object, the modification would have affected every other place and would have been permanent for the life of the session. This seems like a bad idea, especially for the public API of a package.

**Comment by pottedmeat** [comment]
>if any of those places had accidentally modified the object, the modification would have affected every other place

Ah, I misunderstood—I thought this was related to detecting prop mutations. Makes sense!

**Review by nbloomf — APPROVED** [review]
I tested a renewal and an upgrade, and tried some cart-changing actions like changing the plan length and removing an item from the cart. Everything worked as expected.

**Pull request merged**